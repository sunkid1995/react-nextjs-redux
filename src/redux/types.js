export const PREFIXES = {
  AUTH: 'auth/',
  SERVICE_API: 'serviceApi/',
  SHARED: 'shared/',
};

export const INIT_ACTION = { type: '@@redux/INIT' };

export const AUTH = {
  AUTHORIZE: 'auth/AUTHORIZE',
  DEAUTHORIZE: 'auth/DEAUTHORIZE',
};

export const SERVICE_API = {
  AUTH_WITH_EMAIL: 'serviceApi/AUTH_WITH_EMAIL',
  FORGOT_PASSWORD: 'serviceApi/FORGOT_PASSWORD',
  RESET_PASSWORD: 'serviceApi/RESET_PASSWORD',
};

export const SHARED = {

};
