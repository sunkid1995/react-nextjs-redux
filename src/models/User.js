import Joi from 'joi-browser';
import _ from 'lodash';

// Models
import BaseModel, { COMMON_SCHEMAS } from './BaseModel';

const { ID, STRING, STRINGS_ARRAY } = COMMON_SCHEMAS;

const ROLES = {
  MANAGER: 'sales_manager',
  STAFF: 'sales_man',
  USER: 'User',
};

const { MANAGER, STAFF } = ROLES;

export default class User extends BaseModel {
  static get klass() { return 'User'; }
  
  static dataValidator(): Joi.SchemaMap {
    return {
      auth_token: STRING.required(),
      display_name: STRING.required(),
      email: STRING.required(),
      id: ID,
      nickname: STRING.required(),
      picture: STRING,
      roles: STRINGS_ARRAY,
    };
  }

  get isManager() { return _.includes(this.roles, MANAGER); }
  get isStaff() { return _.includes(this.roles, STAFF) && !_.includes(this.roles, MANAGER); }
}
