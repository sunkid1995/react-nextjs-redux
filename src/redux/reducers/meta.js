import { REHYDRATE } from 'redux-persist/lib/constants';

const INITIAL_STATE = {
  rehydrated: false,
};

export default function metaReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === REHYDRATE) {
    const { err } = action;

    /**
     * On server-side, REHYDRATE action will throw an err:
     * Error: No available storage method found
     * Otherwise on client-side, REHYDRATE will be dispatched without error
     * So we check if err is nullable then it's actually rehydrated in client
     */
    if (err == null) return { ...state, rehydrated: true };
  }

  return state;
}
