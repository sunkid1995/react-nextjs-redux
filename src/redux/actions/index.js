import * as auth from './auth';
import * as serviceApi from './serviceApi';
import * as shared from './shared';

export { auth, serviceApi, shared };
