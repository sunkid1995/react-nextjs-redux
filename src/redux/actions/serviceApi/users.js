import { createAction } from 'redux-actions';

// Constants
import { API_CONFIGS } from 'src/constants';

// Models
import User from 'src/models/User';

// Redux
import { SERVICE_API } from 'src/redux/types';

// Utils
// import { buildHeaders, responseInterceptor } from 'src/redux/utils';

const { REQUEST_METHODS: { PATCH, POST } } = API_CONFIGS;

export function authWithEmail({ email, password }) {
  const action = createAction(SERVICE_API.AUTH_WITH_EMAIL);
  const dataKey = 'emailAuth';
  const manifest = { klass: User.klass };

  return dispatch => {
    const request = {
      data: { email, password },
      method: POST,
      url: '/users/login',
    };

    return dispatch(action({ dataKey, manifest, request }));
  };
}

export function forgotPassword(email) {
  const action = createAction(SERVICE_API.FORGOT_PASSWORD);
  const dataKey = 'passwordForgotten';
  const manifest = { reserved: true };

  return dispatch => {
    const request = {
      data: { email },
      method: POST,
      url: '/users/password',
    };

    return dispatch(action({ dataKey, manifest, request }));
  };
}

export function resetPassword({ password, passwordConfirm, token }) {
  const action = createAction(SERVICE_API.RESET_PASSWORD);
  const dataKey = 'passwordReset';
  const manifest = { reserved: true };

  return dispatch => {
    const data = {
      password,
      reset_password_token: token,
      password_confirmation: passwordConfirm,
    };

    const request = {
      data,
      method: PATCH,
      url: '/users/password',
    };

    return dispatch(action({ dataKey, manifest, request }));
  };
}

