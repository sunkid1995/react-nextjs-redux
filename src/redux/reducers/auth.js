// Redux
import { AUTH } from 'src/redux/types';

const INITIAL_STATE = {
  user: undefined,
};

export default function authReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === AUTH.AUTHORIZE) {
    const { payload: user } = action;
    return { ...state, user };
  }

  if (type === AUTH.DEAUTHORIZE) {
    return INITIAL_STATE;
  }

  return state;
}
