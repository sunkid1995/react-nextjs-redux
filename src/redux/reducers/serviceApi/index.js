import _ from 'lodash';

// Constants
import { AUTH, SERVICE_API } from 'src/redux/types';

// Locals
// import blankRequestReducer from './blankRequestReducer';
// import paginatedRequestReducer from './paginatedRequestReducer';
import plainRequestReducer from './plainRequestReducer';

const INITIAL_BLANK_REQUEST = { error: undefined, loading: false };
// const INITIAL_PAGINATED_DATA_REQUEST = { dataPages: {}, totalPages: 1, ...INITIAL_BLANK_REQUEST };
const INITIAL_PLAIN_DATA_REQUEST = { data: {}, ...INITIAL_BLANK_REQUEST };

const INITIAL_STATE = {
  emailAuth: INITIAL_PLAIN_DATA_REQUEST,
  passwordForgotten: INITIAL_PLAIN_DATA_REQUEST,
  passwordReset: INITIAL_PLAIN_DATA_REQUEST,
};

export default function serviceApiReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === AUTH.DEAUTHORIZE) return INITIAL_STATE;

  if (
    _.startsWith(type, SERVICE_API.AUTH_WITH_EMAIL) ||
    _.startsWith(type, SERVICE_API.FORGOT_PASSWORD) ||
    _.startsWith(type, SERVICE_API.RESET_PASSWORD)
  ) return plainRequestReducer(state, action);

  return state;
}
