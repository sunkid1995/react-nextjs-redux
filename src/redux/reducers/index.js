import auth from './auth';
import meta from './meta';
import serviceApi from './serviceApi';
import shared from './shared';

export { auth, meta, serviceApi, shared };
