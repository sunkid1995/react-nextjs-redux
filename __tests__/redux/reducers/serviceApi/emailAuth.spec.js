import { serviceApi } from 'src/redux/reducers';
import { SERVICE_API } from 'src/redux/types';

// Helpers
import buildMockState from '__tests__/helpers/buildMockState';
import { plainDataRequest } from '__tests__/helpers/buildMockState/serviceApi';
import { buildErrorAction, buildSuccessAction } from '__tests__/helpers/mockApiRequests';

// Models
import User from 'src/models/User';

describe('redux/reducers/serviceApi/emailAuth', () => {
  const { serviceApi: mockState } = buildMockState();
  const mockData = User.build(require('src/data/fixtures/user.json'));

  const requestAction = { type: SERVICE_API.AUTH_WITH_EMAIL, payload: { dataKey: 'emailAuth' } };
  const successAction = buildSuccessAction(requestAction, mockData);
  const { errorAction } = buildErrorAction(requestAction, { noResponse: true });

  test('start', () => {
    const expectedState = {
      ...mockState,
      emailAuth: { ...plainDataRequest, loading: true },
    };

    const reduced = serviceApi(mockState, requestAction);
    expect(reduced).toEqual(expectedState);
    expect(reduced).toMatchSnapshot();
  });

  test('success', () => {
    const expectedState = {
      ...mockState,
      emailAuth: { ...plainDataRequest, data: mockData },
    };

    const reduced = serviceApi(mockState, successAction);
    expect(reduced).toEqual(expectedState);
    expect(reduced).toMatchSnapshot();
  });

  test('error', () => {
    const expectedState = {
      ...mockState,
      emailAuth: {
        ...plainDataRequest,
        error: { message: `${SERVICE_API.AUTH_WITH_EMAIL} request failed` },
      },
    };

    const reduced = serviceApi(mockState, errorAction);
    expect(reduced).toEqual(expectedState);
    expect(reduced).toMatchSnapshot();
  });

  test('refresh', () => {
    const currentState = {
      ...mockState,
      emailAuth: { ...plainDataRequest, data: mockData },
    };

    const expectedState = {
      ...mockState,
      emailAuth: { ...plainDataRequest, data: mockData, loading: true },
    };

    const reduced = serviceApi(currentState, requestAction);
    expect(reduced).toEqual(expectedState);
    expect(reduced).toMatchSnapshot();
  });
});
