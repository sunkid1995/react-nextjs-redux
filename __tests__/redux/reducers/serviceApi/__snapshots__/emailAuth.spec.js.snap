// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux/reducers/serviceApi/emailAuth error 1`] = `
Object {
  "emailAuth": Object {
    "data": undefined,
    "error": Object {
      "message": "serviceApi/AUTH_WITH_EMAIL request failed",
      "response": undefined,
    },
    "loading": false,
  },
  "passwordForgotten": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
  "passwordReset": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
}
`;

exports[`redux/reducers/serviceApi/emailAuth refresh 1`] = `
Object {
  "emailAuth": Object {
    "data": User {
      "auth_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJlbWFpbCI6InRoYW5obnhAb25hY2xvdmVyLmNvbSIsInVpZCI6IjgzOWU2OGJkLTJjN2QtNDkzMy1hMmI3LWYxZDM0MDkyODI0NiIsImlhdCI6MTUyMDg0MjA3MCwiZXhwIjoxNTIxNDQ2ODcwLCJqdGkiOiJiMDlhYjM4NS1iZWZiLTRiNzYtODQyZC0xODQ1ZWU2MGJhNGEifQ.zUxEMCY2zYj6y4jw6-sJAHig-hPzDY_l_MoCT4XfbhI",
      "display_name": "ThanhNX",
      "email": "thanhnx@onaclover.com",
      "id": 4,
      "nickname": "ThanhNX",
      "roles": Array [
        "user",
        "sales_manager",
        "sales_man",
      ],
    },
    "error": undefined,
    "loading": true,
  },
  "passwordForgotten": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
  "passwordReset": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
}
`;

exports[`redux/reducers/serviceApi/emailAuth start 1`] = `
Object {
  "emailAuth": Object {
    "data": undefined,
    "error": undefined,
    "loading": true,
  },
  "passwordForgotten": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
  "passwordReset": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
}
`;

exports[`redux/reducers/serviceApi/emailAuth success 1`] = `
Object {
  "emailAuth": Object {
    "data": User {
      "auth_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJlbWFpbCI6InRoYW5obnhAb25hY2xvdmVyLmNvbSIsInVpZCI6IjgzOWU2OGJkLTJjN2QtNDkzMy1hMmI3LWYxZDM0MDkyODI0NiIsImlhdCI6MTUyMDg0MjA3MCwiZXhwIjoxNTIxNDQ2ODcwLCJqdGkiOiJiMDlhYjM4NS1iZWZiLTRiNzYtODQyZC0xODQ1ZWU2MGJhNGEifQ.zUxEMCY2zYj6y4jw6-sJAHig-hPzDY_l_MoCT4XfbhI",
      "display_name": "ThanhNX",
      "email": "thanhnx@onaclover.com",
      "id": 4,
      "nickname": "ThanhNX",
      "roles": Array [
        "user",
        "sales_manager",
        "sales_man",
      ],
    },
    "error": undefined,
    "loading": false,
  },
  "passwordForgotten": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
  "passwordReset": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
}
`;
